#7507 degrades the Pydantic model system 🥲

Sorry @agoscinski. I'm afraid #7507 regressed my REST API behavior and is now failing my `test_get_node_schema_by_type` test
```python
    @pytest.mark.parametrize(
        'which, model, name',
        [
            ['read', orm.Int.ReadModel, 'AttributesModel'],
            ['write', orm.Int.WriteModel, 'AttributesWriteModel'],
        ],
    )
    def test_get_node_schema_by_type(client: TestClient, which: str, model: type[BaseModel], name: str):
        """Test get schema for nodes by valid type."""
        response = client.get(f'/nodes/schema?type=data.core.int.Int.&which={which}')
        assert response.status_code == 200
        result = response.json()
        assert 'properties' in result
        assert sorted(result['properties'].keys()) == sorted(model.model_fields.keys())
        assert 'attributes' in result['properties']
        attributes = result['properties']['attributes']
        assert '$ref' in attributes
        assert attributes['$ref'].endswith(name)
        assert '$defs' in result
        assert name in result['$defs']
>       assert result['$defs'][name]['title'] == f'Int{name}'
E       AssertionError: assert 'IntAttributesModel' == 'IntAttributesWriteModel'
E         - IntAttributesWriteModel
E         ?              -----
E         + IntAttributesModel
```
When generating a "write" version of a `ReadModel`, we change the name of the model, as well as it's nested `AttributesModel`, to both having `WriteModel` as a suffix. After dynamically changing it, the model needed a rebuild. Without it, this is not reflected. Missed it when reviewing this PR 🥲

Maybe there's a compromise between performance and correctness that we can arrive at 🙏

--- a/src/aiida/orm/entities.py
+++ b/src/aiida/orm/entities.py
@@ -386,7 +386,7 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
             except TypeError:
                 return Any | None
 
-        model_fields: dict[str, Any] = {
+        model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {
             key: (field.annotation, deepcopy(field)) for key, field in cls.WriteModel.model_fields.items()
         }
 
@@ -586,7 +586,7 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
                 bases.append(OrmFieldsAsModelDump)  # write models should inherit the override
             bases.append(base_cls)
 
-            model_fields: dict[str, Any] = {
+            model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {
                 key: copy_model_field(field)
                 for key, field in model_cls.model_fields.items()
                 if not get_metadata(field, 'read_only', False)

--- a/src/aiida/orm/entities.py
+++ b/src/aiida/orm/entities.py
@@ -556,6 +556,24 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
             :return: The derived creation model class.
             """
 
+            from pydantic._internal._decorators import PydanticDescriptorProxy
+
+            def copy_decorator(model_cls: type[OrmModel], name: str, decorator: Any) -> PydanticDescriptorProxy:
+                """Reconstruct a Pydantic decorator for installation on a new model."""
+
+                for parent in model_cls.__mro__:
+                    if name in parent.__dict__:
+                        wrapped = parent.__dict__[name]
+                        break
+                else:
+                    raise RuntimeError(f'Could not locate decorated attribute {name!r} on {model_cls.__qualname__}')
+
+                return PydanticDescriptorProxy(
+                    wrapped=wrapped,
+                    decorator_info=deepcopy(decorator.info),
+                    shim=decorator.shim,
+                )
+
             def copy_model_field(field: pdt.fields.FieldInfo) -> tuple[Any, pdt.fields.FieldInfo]:
                 """Copy a model field, replacing any nested read models with their write model equivalent."""
                 annotation = field.annotation
@@ -574,32 +592,30 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
                 if not get_metadata(field, 'read_only', False)
             }
 
-            serializers = {
-                key: deepcopy(serializer)
-                for key, serializer in model_cls.__pydantic_decorators__.field_serializers.items()
-                if all(serializer_key in model_fields for serializer_key in serializer.info.fields)
-            }
+            decorators: dict[str, Any] = {}
 
-            validators = {
-                key: deepcopy(validator)
-                for key, validator in model_cls.__pydantic_decorators__.field_validators.items()
-                if all(validator_key in model_fields for validator_key in validator.info.fields)
-            }
+            for registry in (
+                model_cls.__pydantic_decorators__.field_validators,
+                model_cls.__pydantic_decorators__.field_serializers,
+            ):
+                for name, decorator in registry.items():
+                    # Only retain decorators for the remaining non-read-only fields
+                    if all(field in model_fields for field in decorator.info.fields):
+                        decorators[name] = copy_decorator(model_cls, name, decorator)
 
             name = model_cls.__name__.replace(suffix, 'WriteModel')
             WriteModel = cast(  # noqa: N806
                 type[OrmModel],
                 pdt.create_model(
                     name,
+                    __config__=deepcopy(model_cls.model_config),
                     __base__=tuple(bases),
                     __module__=model_cls.__module__,
+                    __qualname__=model_cls.__qualname__.replace(suffix, 'WriteModel'),
+                    __validators__=decorators,
                     **model_fields,
                 ),
             )
-            WriteModel.__qualname__ = model_cls.__qualname__.replace(suffix, 'WriteModel')
-            WriteModel.__pydantic_decorators__.field_serializers = serializers
-            WriteModel.__pydantic_decorators__.field_validators = validators
-            WriteModel.model_config = deepcopy(model_cls.model_config)
 
             return WriteModel
 

--- a/src/aiida/orm/entities.py
+++ b/src/aiida/orm/entities.py
@@ -408,7 +408,6 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
             type[OrmModel],
             pdt.create_model(
                 'Model',
-                __config__=deepcopy(cls.ReadModel.model_config),
                 __base__=OrmModel,
                 __module__=cls.ReadModel.__module__,
                 __qualname__=f'{cast(Any, cls).__name__}.Model',

--- a/src/aiida/orm/entities.py
+++ b/src/aiida/orm/entities.py
@@ -408,13 +408,13 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
             type[OrmModel],
             pdt.create_model(
                 'Model',
+                __config__=deepcopy(cls.ReadModel.model_config),
                 __base__=OrmModel,
                 __module__=cls.ReadModel.__module__,
+                __qualname__=f'{cast(Any, cls).__name__}.Model',
                 **model_fields,
             ),
         )
-        model.__qualname__ = f'{cast(Any, cls).__name__}.Model'
-        model.model_config = deepcopy(cls.ReadModel.model_config)
 
         cls._COMPAT_MODEL = model
 

--- a/src/aiida/orm/entities.py
+++ b/src/aiida/orm/entities.py
@@ -556,24 +556,6 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
             :return: The derived creation model class.
             """
 
-            from pydantic._internal._decorators import PydanticDescriptorProxy
-
-            def copy_decorator(model_cls: type[OrmModel], name: str, decorator: Any) -> PydanticDescriptorProxy:
-                """Reconstruct a Pydantic decorator for installation on a new model."""
-
-                for parent in model_cls.__mro__:
-                    if name in parent.__dict__:
-                        wrapped = parent.__dict__[name]
-                        break
-                else:
-                    raise RuntimeError(f'Could not locate decorated attribute {name!r} on {model_cls.__qualname__}')
-
-                return PydanticDescriptorProxy(
-                    wrapped=wrapped,
-                    decorator_info=deepcopy(decorator.info),
-                    shim=decorator.shim,
-                )
-
             def copy_model_field(field: pdt.fields.FieldInfo) -> tuple[Any, pdt.fields.FieldInfo]:
                 """Copy a model field, replacing any nested read models with their write model equivalent."""
                 annotation = field.annotation
@@ -592,16 +574,17 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
                 if not get_metadata(field, 'read_only', False)
             }
 
-            decorators: dict[str, Any] = {}
+            serializers = {
+                key: deepcopy(serializer)
+                for key, serializer in model_cls.__pydantic_decorators__.field_serializers.items()
+                if all(serializer_key in model_fields for serializer_key in serializer.info.fields)
+            }
 
-            for registry in (
-                model_cls.__pydantic_decorators__.field_validators,
-                model_cls.__pydantic_decorators__.field_serializers,
-            ):
-                for name, decorator in registry.items():
-                    # Only retain decorators for the remaining non-read-only fields
-                    if all(field in model_fields for field in decorator.info.fields):
-                        decorators[name] = copy_decorator(model_cls, name, decorator)
+            validators = {
+                key: deepcopy(validator)
+                for key, validator in model_cls.__pydantic_decorators__.field_validators.items()
+                if all(validator_key in model_fields for validator_key in validator.info.fields)
+            }
 
             name = model_cls.__name__.replace(suffix, 'WriteModel')
             WriteModel = cast(  # noqa: N806
@@ -612,10 +595,11 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
                     __base__=tuple(bases),
                     __module__=model_cls.__module__,
                     __qualname__=model_cls.__qualname__.replace(suffix, 'WriteModel'),
-                    __validators__=decorators,
                     **model_fields,
                 ),
             )
+            WriteModel.__pydantic_decorators__.field_serializers = serializers
+            WriteModel.__pydantic_decorators__.field_validators = validators
 
             return WriteModel
 

--- a/src/aiida/orm/entities.py
+++ b/src/aiida/orm/entities.py
@@ -386,7 +386,7 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
             except TypeError:
                 return Any | None
 
-        model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {
+        model_fields: dict[str, Any] = {
             key: (field.annotation, deepcopy(field)) for key, field in cls.WriteModel.model_fields.items()
         }
 
@@ -586,7 +586,7 @@ class Entity(abc.ABC, Generic[BackendEntityType, CollectionType]):
                 bases.append(OrmFieldsAsModelDump)  # write models should inherit the override
             bases.append(base_cls)
 
-            model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {
+            model_fields: dict[str, Any] = {
                 key: copy_model_field(field)
                 for key, field in model_cls.model_fields.items()
                 if not get_metadata(field, 'read_only', False)

--- a/src/aiida/orm/nodes/data/code/abstract.py
+++ b/src/aiida/orm/nodes/data/code/abstract.py
@@ -202,7 +202,7 @@ class AbstractCode(Data, metaclass=abc.ABCMeta):
     @classmethod
     def _patch_cli_model(cls):
         """Patch `CliModel` by synthesizing it from the base and constructor models."""
-        model_fields: dict[str, t.Any] = {
+        model_fields: dict[str, tuple[t.Any, pdt.fields.FieldInfo]] = {
             'label': (
                 cls.BaseNodeModel.model_fields['label'].annotation,
                 deepcopy(cls.BaseNodeModel.model_fields['label']),

--- a/src/aiida/orm/nodes/data/code/abstract.py
+++ b/src/aiida/orm/nodes/data/code/abstract.py
@@ -220,7 +220,7 @@ class AbstractCode(Data, metaclass=abc.ABCMeta):
             type[OrmModel],
             pdt.create_model(
                 'CliModel',
-                __config__=deepcopy(cls.ConstructorArgsModel.model_config) | {'arbitrary_types_allowed': True},
+                __config__=deepcopy(cls.ConstructorArgsModel.model_config),
                 __base__=OrmModel,
                 __module__=cls.__module__,
                 __qualname__=f'{cls.__name__}.CliModel',

--- a/src/aiida/orm/nodes/data/code/abstract.py
+++ b/src/aiida/orm/nodes/data/code/abstract.py
@@ -220,7 +220,6 @@ class AbstractCode(Data, metaclass=abc.ABCMeta):
             type[OrmModel],
             pdt.create_model(
                 'CliModel',
-                __config__=deepcopy(cls.ConstructorArgsModel.model_config),
                 __base__=OrmModel,
                 __module__=cls.__module__,
                 __qualname__=f'{cls.__name__}.CliModel',

--- a/src/aiida/orm/nodes/data/code/abstract.py
+++ b/src/aiida/orm/nodes/data/code/abstract.py
@@ -202,7 +202,7 @@ class AbstractCode(Data, metaclass=abc.ABCMeta):
     @classmethod
     def _patch_cli_model(cls):
         """Patch `CliModel` by synthesizing it from the base and constructor models."""
-        model_fields: dict[str, tuple[t.Any, pdt.fields.FieldInfo]] = {
+        model_fields: dict[str, t.Any] = {
             'label': (
                 cls.BaseNodeModel.model_fields['label'].annotation,
                 deepcopy(cls.BaseNodeModel.model_fields['label']),

--- a/src/aiida/orm/nodes/data/code/abstract.py
+++ b/src/aiida/orm/nodes/data/code/abstract.py
@@ -220,13 +220,13 @@ class AbstractCode(Data, metaclass=abc.ABCMeta):
             type[OrmModel],
             pdt.create_model(
                 'CliModel',
+                __config__=deepcopy(cls.ConstructorArgsModel.model_config) | {'arbitrary_types_allowed': True},
                 __base__=OrmModel,
                 __module__=cls.__module__,
+                __qualname__=f'{cls.__name__}.CliModel',
                 **model_fields,
             ),
         )
-        CliModel.__qualname__ = f'{cls.__name__}.CliModel'
-        CliModel.model_config['arbitrary_types_allowed'] = True
         cls._CliModel = CliModel
 
     @abc.abstractmethod

--- a/src/aiida/orm/nodes/node.py
+++ b/src/aiida/orm/nodes/node.py
@@ -448,7 +448,6 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
             type[OrmModel],
             pdt.create_model(
                 'Model',
-                __config__=deepcopy(cls.ReadModel.model_config),
                 __base__=OrmModel,
                 __module__=cls.ReadModel.__module__,
                 __qualname__=f'{cast(Any, cls).__name__}.Model',

--- a/src/aiida/orm/nodes/node.py
+++ b/src/aiida/orm/nodes/node.py
@@ -448,13 +448,13 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
             type[OrmModel],
             pdt.create_model(
                 'Model',
+                __config__=deepcopy(cls.ReadModel.model_config),
                 __base__=OrmModel,
                 __module__=cls.ReadModel.__module__,
+                __qualname__=f'{cast(Any, cls).__name__}.Model',
                 **cast(dict[str, Any], model_fields),
             ),
         )
-        model.__qualname__ = f'{cast(Any, cls).__name__}.Model'
-        model.model_config = deepcopy(cls.ReadModel.model_config)
 
         cls._COMPAT_MODEL = model
 
@@ -1172,9 +1172,9 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
                     'AttributesModel',
                     __base__=cls.AttributesModel,
                     __module__=cls.__module__,
+                    __qualname__=f'{cls.__name__}.AttributesModel',
                 ),
             )
-            AttributesModel.__qualname__ = f'{cls.__name__}.AttributesModel'
             cls.AttributesModel = AttributesModel  # type: ignore[misc]
 
     @classmethod
@@ -1230,10 +1230,10 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
                 'ReadModel',
                 __base__=BaseReadModel,
                 __module__=cls.__module__,
+                __qualname__=f'{cls.__name__}.ReadModel',
                 **model_fields,
             ),
         )
-        ReadModel.__qualname__ = f'{cls.__name__}.ReadModel'
 
         cls.ReadModel = ReadModel  # type: ignore[misc]
 
@@ -1258,10 +1258,10 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
                 'ConstructorModel',
                 __base__=cls.BaseNodeModel,
                 __module__=cls.__module__,
+                __qualname__=f'{cls.__name__}.ConstructorModel',
                 **model_fields,
             ),
         )
-        ConstructorModel.__qualname__ = f'{cls.__name__}.ConstructorModel'
 
         cls._ConstructorModel = ConstructorModel
 

--- a/src/aiida/orm/nodes/node.py
+++ b/src/aiida/orm/nodes/node.py
@@ -395,7 +395,7 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
 
             return annotation, merged_field
 
-        model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {}
+        model_fields: dict[str, Any] = {}
 
         for key, field in cls.WriteModel.model_fields.items():
             if key in {'attributes', 'node_type'}:
@@ -452,7 +452,7 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
                 __base__=OrmModel,
                 __module__=cls.ReadModel.__module__,
                 __qualname__=f'{cast(Any, cls).__name__}.Model',
-                **cast(dict[str, Any], model_fields),
+                **model_fields,
             ),
         )
 
@@ -1198,7 +1198,7 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
 
         BaseReadModel: type[Node.ReadModel] = cls.ReadModel  # noqa: N806
 
-        model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {}
+        model_fields: dict[str, Any] = {}
 
         if 'ReadModel' in cls.__dict__:
             # TODO ideally we should do this check with issubclass, but we can't import
@@ -1248,7 +1248,7 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
             description='The constructor arguments.',
             write_only=True,
         )
-        model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {
+        model_fields: dict[str, Any] = {
             'args': (cls.ConstructorArgsModel, args_field),
             'node_type': cls._get_patched_node_type_field(),
         }

--- a/src/aiida/orm/nodes/node.py
+++ b/src/aiida/orm/nodes/node.py
@@ -1217,7 +1217,7 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
             # For exceptions that override `ReadModel`, we need to copy the overridden fields.
             # We don't know a priori which fields are overridden, so we copy all.
             BaseReadModel = cls.ReadModel.__bases__[0]  # noqa: N806
-            model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {
+            model_fields = {
                 key: (field.annotation, deepcopy(field)) for key, field in cls.ReadModel.model_fields.items()
             }
 

--- a/src/aiida/orm/nodes/node.py
+++ b/src/aiida/orm/nodes/node.py
@@ -1197,7 +1197,8 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
         """
 
         BaseReadModel: type[Node.ReadModel] = cls.ReadModel  # noqa: N806
-        model_fields: dict[str, Any] = {}
+
+        model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {}
 
         if 'ReadModel' in cls.__dict__:
             # TODO ideally we should do this check with issubclass, but we can't import
@@ -1216,7 +1217,7 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
             # For exceptions that override `ReadModel`, we need to copy the overridden fields.
             # We don't know a priori which fields are overridden, so we copy all.
             BaseReadModel = cls.ReadModel.__bases__[0]  # noqa: N806
-            model_fields = {
+            model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {
                 key: (field.annotation, deepcopy(field)) for key, field in cls.ReadModel.model_fields.items()
             }
 
@@ -1247,7 +1248,7 @@ class Node(Entity['BackendNode', NodeCollection['Node']], metaclass=AbstractNode
             description='The constructor arguments.',
             write_only=True,
         )
-        model_fields: dict[str, Any] = {
+        model_fields: dict[str, tuple[Any, pdt.fields.FieldInfo]] = {
             'args': (cls.ConstructorArgsModel, args_field),
             'node_type': cls._get_patched_node_type_field(),
         }

--- a/src/aiida/orm/pydantic.py
+++ b/src/aiida/orm/pydantic.py
@@ -75,7 +75,7 @@ class OrmModel(AiiDABaseModel):
         except ValueError as exception:
             raise ValueError(f"expected 'OrmClass.ModelName' format, got '{cls.__qualname__}'") from exception
 
-        model_fields: dict[str, tuple[t.Any, pdt.fields.FieldInfo]] = {}
+        model_fields: dict[str, t.Any] = {}
         for key, field in cls.model_fields.items():
             annotation = field.annotation
             field_copy = deepcopy(field)

--- a/src/aiida/orm/pydantic.py
+++ b/src/aiida/orm/pydantic.py
@@ -1,8 +1,8 @@
 from __future__ import annotations
 
-from copy import deepcopy
 import datetime
 import typing as t
+from copy import deepcopy
 
 import pydantic as pdt
 from pydantic import create_model
@@ -82,18 +82,21 @@ class OrmModel(AiiDABaseModel):
                     field_copy.default_factory = None
             model_fields[key] = (field_copy.annotation, field_copy)
 
-        MinimalModel = create_model(  # noqa: N806
-            f'Minimal{model_name}',
-            __config__=deepcopy(cls.model_config) | {'extra': 'ignore'},
-            __base__=OrmModel,
-            __module__=cls.__module__,
-            __qualname__=f'{orm_class_name}.Minimal{model_name}',
-            **model_fields,
+        MinimalModel = t.cast(  # noqa: N806
+            type[OrmModel],
+            create_model(
+                f'Minimal{model_name}',
+                __config__=deepcopy(cls.model_config) | {'extra': 'ignore'},
+                __base__=OrmModel,
+                __module__=cls.__module__,
+                __qualname__=f'{orm_class_name}.Minimal{model_name}',
+                **model_fields,
+            ),
         )
 
         # Make subsequent calls idempotent for this specific class and the derived model
-        cls._AIIDA_MINIMAL_MODEL = MinimalModel  # type: ignore[attr-defined]
-        MinimalModel._AIIDA_MINIMAL_MODEL = MinimalModel  # type: ignore[attr-defined]
+        cls._AIIDA_MINIMAL_MODEL = MinimalModel
+        MinimalModel._AIIDA_MINIMAL_MODEL = MinimalModel
 
         return MinimalModel
 

--- a/src/aiida/orm/pydantic.py
+++ b/src/aiida/orm/pydantic.py
@@ -20,6 +20,8 @@ __all__ = ('OrmModel',)
 class OrmModel(AiiDABaseModel):
     """Base class for Read/Write/Attributes models."""
 
+    _AIIDA_MINIMAL_MODEL: type[OrmModel] | None = None
+
     model_config = pdt.ConfigDict(
         extra='forbid',
         json_encoders={
@@ -61,9 +63,13 @@ class OrmModel(AiiDABaseModel):
     @classmethod
     def _as_minimal_model(cls: type[OrmModel]) -> type[OrmModel]:
         """Return a derived model class excluding fields marked as "may_be_large"."""
-        cached = cls.__dict__.get('_AIIDA_MINIMAL_MODEL')
+        if cls.__name__.startswith('Minimal'):
+            return cls
+
+        cached = cls._AIIDA_MINIMAL_MODEL
         if isinstance(cached, type) and issubclass(cached, OrmModel):
             return cached
+
         try:
             orm_class_name, model_name = cls.__qualname__.split('.')
         except ValueError as exception:
@@ -94,9 +100,7 @@ class OrmModel(AiiDABaseModel):
             ),
         )
 
-        # Make subsequent calls idempotent for this specific class and the derived model
-        cls._AIIDA_MINIMAL_MODEL = MinimalModel
-        MinimalModel._AIIDA_MINIMAL_MODEL = MinimalModel
+        cls._AIIDA_MINIMAL_MODEL = MinimalModel  # cache the derived model on the original class
 
         return MinimalModel
 

--- a/src/aiida/orm/pydantic.py
+++ b/src/aiida/orm/pydantic.py
@@ -72,14 +72,15 @@ class OrmModel(AiiDABaseModel):
         model_fields: dict[str, tuple[t.Any, pdt.fields.FieldInfo]] = {}
         for key, field in cls.model_fields.items():
             annotation = field.annotation
+            field_copy = deepcopy(field)
             if get_metadata(field, 'may_be_large'):
                 continue
             if isinstance(annotation, type) and issubclass(annotation, OrmModel):
                 sub_minimal_model = annotation._as_minimal_model()
-                field.annotation = sub_minimal_model
+                field_copy.annotation = sub_minimal_model
                 if any(f.is_required() for f in sub_minimal_model.model_fields.values()):
-                    field.default_factory = None
-            model_fields[key] = (field.annotation, deepcopy(field))
+                    field_copy.default_factory = None
+            model_fields[key] = (field_copy.annotation, field_copy)
 
         MinimalModel = create_model(  # noqa: N806
             f'Minimal{model_name}',

--- a/src/aiida/orm/pydantic.py
+++ b/src/aiida/orm/pydantic.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+from copy import deepcopy
 import datetime
 import typing as t
 
@@ -67,14 +68,8 @@ class OrmModel(AiiDABaseModel):
             orm_class_name, model_name = cls.__qualname__.split('.')
         except ValueError as exception:
             raise ValueError(f"expected 'OrmClass.ModelName' format, got '{cls.__qualname__}'") from exception
-        MinimalModel = create_model(  # noqa: N806
-            f'Minimal{model_name}',
-            __base__=OrmModel,
-            __module__=cls.__module__,
-        )
-        MinimalModel.__qualname__ = f'{orm_class_name}.Minimal{model_name}'
-        MinimalModel.model_config['extra'] = 'ignore'
 
+        model_fields: dict[str, tuple[t.Any, pdt.fields.FieldInfo]] = {}
         for key, field in cls.model_fields.items():
             annotation = field.annotation
             if get_metadata(field, 'may_be_large'):
@@ -84,7 +79,16 @@ class OrmModel(AiiDABaseModel):
                 field.annotation = sub_minimal_model
                 if any(f.is_required() for f in sub_minimal_model.model_fields.values()):
                     field.default_factory = None
-            MinimalModel.model_fields[key] = field
+            model_fields[key] = (field.annotation, deepcopy(field))
+
+        MinimalModel = create_model(  # noqa: N806
+            f'Minimal{model_name}',
+            __config__=deepcopy(cls.model_config) | {'extra': 'ignore'},
+            __base__=OrmModel,
+            __module__=cls.__module__,
+            __qualname__=f'{orm_class_name}.Minimal{model_name}',
+            **model_fields,
+        )
 
         # Make subsequent calls idempotent for this specific class and the derived model
         cls._AIIDA_MINIMAL_MODEL = MinimalModel  # type: ignore[attr-defined]

--- a/src/aiida/orm/pydantic.py
+++ b/src/aiida/orm/pydantic.py
@@ -5,7 +5,6 @@ import typing as t
 from copy import deepcopy
 
 import pydantic as pdt
-from pydantic import create_model
 from pydantic_core import PydanticUndefined
 
 from aiida.common.exceptions import EntryPointError, NotExistent
@@ -90,7 +89,7 @@ class OrmModel(AiiDABaseModel):
 
         MinimalModel = t.cast(  # noqa: N806
             type[OrmModel],
-            create_model(
+            pdt.create_model(
                 f'Minimal{model_name}',
                 __config__=deepcopy(cls.model_config) | {'extra': 'ignore'},
                 __base__=OrmModel,

--- a/tests/orm/models/test_models.py
+++ b/tests/orm/models/test_models.py
@@ -540,7 +540,7 @@ def test_generated_orm_model_setup_defers_pydantic_rebuild(monkeypatch):
 
     model = TestData.WriteModel(node_type=TestData.class_node_type, attributes={'value': '1'})
     assert model.attributes.value == 1
-    assert TestData.ReadModel.model_json_schema()['title'] == 'TestDataReadModel'
+    assert TestData.ReadModel.model_config.get('title') == 'TestDataReadModel'
 
 
 @pytest.mark.parametrize(
@@ -553,10 +553,10 @@ def test_model_overrides(required_arguments: RequiredEntityArguments):
     name = cls.__name__
 
     assert cls.ReadModel.__qualname__ == f'{name}.ReadModel'
-    assert cls.ReadModel.model_json_schema()['title'] == f'{name}ReadModel'
+    assert cls.ReadModel.model_config.get('title') == f'{name}ReadModel'
 
     assert cls.WriteModel.__qualname__ == f'{name}.WriteModel'
-    assert cls.WriteModel.model_json_schema()['title'] == f'{name}WriteModel'
+    assert cls.WriteModel.model_config.get('title') == f'{name}WriteModel'
 
 
 def _clean_and_sort(dictionary: dict) -> dict:
@@ -642,11 +642,11 @@ def test_node_attributes_model_overrides(required_arguments: RequiredNodeArgumen
     AttributesModel = cls.ReadModel.model_fields['attributes'].annotation  # noqa: N806
     assert AttributesModel is cls.AttributesModel
     assert AttributesModel.__qualname__ == f'{name}.AttributesModel'
-    assert AttributesModel.model_json_schema()['title'] == f'{name}AttributesModel'
+    assert AttributesModel.model_config.get('title') == f'{name}AttributesModel'
 
     AttributesWriteModel = cls.WriteModel.model_fields['attributes'].annotation  # noqa: N806
     assert AttributesWriteModel.__qualname__ == f'{name}.AttributesWriteModel'
-    assert AttributesWriteModel.model_json_schema()['title'] == f'{name}AttributesWriteModel'
+    assert AttributesWriteModel.model_config.get('title') == f'{name}AttributesWriteModel'
 
 
 def _validate_value(value):

